client/updates/apt.go

```
@@ -0,0 +1,151 @@
+package updates
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type AptPackageManager struct {
+	runner                 Runner
+	descriptionRegex       *regexp.Regexp
+	rebootRequiredFilename string
+	detectCmd              []string
+	updateCacheCmd         []string
+	getSummariesCmd        []string
+	getCountsCmd           []string
+}
+
+func NewAptPackageManager() *AptPackageManager {
+	return &AptPackageManager{
+		runner:                 &RunnerImpl{},
+		descriptionRegex:       regexp.MustCompile(`\((.*?)\)`),
+		rebootRequiredFilename: "/var/run/reboot-required",
+		detectCmd:              []string{"apt-get", "help"},
+		updateCacheCmd:         []string{"sudo", "-n", "apt-get", "update", "-o", "Debug::NoLocking=true"},
+		getSummariesCmd:        []string{"apt-get", "-s", "-o", "Debug::NoLocking=true", "upgrade"},
+		getCountsCmd:           []string{"/usr/lib/update-notifier/apt-check"},
+	}
+}
+
+func (p *AptPackageManager) Detect() bool {
+	_, err := p.runner.Run(p.detectCmd...)
+	return err == nil
+}
+
+func (p *AptPackageManager) GetUpdatesStatus() *Status {
+	err := p.updatePackageCache()
+	if err != nil {
+		return &Status{
+			Error: err.Error(),
+		}
+	}
+
+	summaries, err := p.getSummaries()
+	if err != nil {
+		return &Status{
+			Error: err.Error(),
+		}
+	}
+
+	rebootPending, err := p.checkRebootRequired()
+	if err != nil {
+		return &Status{
+			Error: err.Error(),
+		}
+	}
+
+	// Get the number of available updates and the number of security updates
+	// from the output of the command, if is available. Ubuntu does some magic to calculate the updates.
+	// As this number is displayed on the system, we give it preference over our own sums.
+	// If not available, we count the summaries.
+	available, security, err := p.getCounts()
+	if err != nil {
```

looks like error is ignored completely, what if the error doesn't indicate the missing counts but contains some internal go error like wrong inputs or code problems? I think we should not silently ignore any errors from our custom method but provide some logic to analyse the error type and detect the difference between errors coming from code from errors coming from external tools.

---

You're right, added a check if it's a command error than we ignore it, but if it's our code's error than we log it. I don't want to fail, since we can still count manually, but you're right we should log the error.